### PR TITLE
2D Effect Snow Fall

### PR DESCRIPTION
**Pull request by Brandon502**
Effect that simulates snow/sand falling and melting. 

Uses getBitValue and setBitValue from Game of Life to reduce memory usage to 1 bit per led. 

Sliders:
- Effect Speed:  0 to 60 updates per second
- Spawn Chance: 0% to ~39% chance 
- Despawn Chance: 0% to ~38% chance.  255 = 100% chance
- Blur: Fade previous snow location
- Sway Chance: 0% to 100% chance of swaying while free falling

Check boxes:
- Use Palette: Uses palette instead of default snow colors
- Overlay: Allows background effect to color the snow
- Prevent Overflow: If snow reaches the top the bottom opens and empties current snow

I've optimized this as much as I can, easily hits 60fps on 32x32. Likely can still be improved. 

Examples on a 32x32 split into 2 segments using different slider/option configs.

https://github.com/user-attachments/assets/9ff35011-7358-4ec9-9716-f558ba11b5dc

https://github.com/user-attachments/assets/2649ffc0-a8ad-4576-9332-b1638cf343e6

https://github.com/user-attachments/assets/4240b83c-8d14-43e3-944c-164153ca9c7d

Naming and default values can be changed however you'd like. Started out as snow, but palette colors look really nice. Could likely be modified to have an audio reactive version as well.

### CODE REVIEW

**Comment by softhack007** [comment]
Hi @Brandon502, nice effect, thank you 😃

I'm a bit busy right now - just tested your effect quickly on 64x64 and it looks very good. 
I can take a closer look at the code in the next weeks.

cheers

File: wled00/FX.cpp [hunk 1]
```diff
@@ -5366,6 +5366,117 @@ uint16_t mode_2Dgameoflife(void) { // Written by Ewoud Wijma, inspired by https:
...
+  uint8_t speed = map(SEGMENT.speed, 0, 255, 0, 60);                      // Updates per second
+  if (!speed || strip.now - SEGENV.step < 1000 / speed) return FRAMETIME; // Not enough time passed
+
+  uint8_t blur = map(SEGMENT.custom2, 0, 255, 255, 0);
+  uint8_t sway = SEGMENT.custom3;
+
+  // Despawn snow
+  bool overflow = SEGENV.aux0 && SEGMENT.check3;
+  int despawnChance = SEGMENT.custom1 == 255 ? 256 : map(SEGMENT.custom1, 0, 255, 0, 100); // 255 goes to 256, allows always despawn
+  int y = rows - 1;
```
Comment (softhack007) [inline wled00/FX.cpp]: "y" causes a warning later (line 5423) about shadowing a local variable.
I'd suggest you name this one "lowY"

File: wled00/FX.cpp [hunk 2]
```diff
@@ -5366,6 +5366,117 @@ uint16_t mode_2Dgameoflife(void) { // Written by Ewoud Wijma, inspired by https:
...
+  for (int x = 0; x < cols; x++) {
+    if (overflow || random8() < despawnChance) setBitValue(grid, y * cols + x, 0);
+    if (overlay  || getBitValue(grid, y * cols + x)) continue; // Skip drawing if inverted overlay or snow
+    SEGMENT.blendPixelColorXY(x, y, bgColor, blur);
+  }
+  if (SEGENV.aux0) --SEGENV.aux0; // Decrease overflow 
+
+  // Precompute shuffled indices, helps randomize snow movement
+  int shuffledIndices[cols];
```
Comment (softhack007) [inline wled00/FX.cpp]: this variable-sized array is allocated from the heap. You could change the type to `uint16_t` to reduce heap pressure a bit.

File: wled00/FX.h [hunk 3]
```diff
@@ -349,8 +349,9 @@ bool strip_uses_global_leds(void);              // WLEDMM implemented in FX_fcn.
 // #define FX_MODE_PALETTE_AR             193 // WLED-SR audioreactive palette
 #define FX_MODE_FIREWORKS_AR           194 // WLED-SR audioreactive fireworks
 #define FX_MODE_GEQLASER                  195 // WLED-MM GEQ Laser
+#define FX_MODE_2DSNOWFALL             196 // WLED-MM Snowfall
```
Comment (softhack007) [inline wled00/FX.h]: @troyhacks already claimed 196 for his new effect - next available one is 197

**Review by softhack007 — APPROVED** [review]
just some minor comments from my side (see above).
Just see if you want to fix them, then increase the FX number.

Thanks 👍

**Comment by Brandon502** [comment]
> just some minor comments from my side (see above). Just see if you want to fix them, then increase the FX number.
> 
> Thanks 👍

Thank you for reviewing. Made all the recommended changes.

**Pull request merged**